org/diceresearch/nebula/orchestrator.py

```diff
@@ -2,140 +2,141 @@
-            logging.exception("Error {0} with json {1}".format(e, EVIDENCE_RETRIEVAL_RESULT))
-            databasemanager.update_step(settings.results_table_name, settings.status, settings.error, identifier)
-            databasemanager.update_step(settings.results_table_name, settings.error_msg, str(e), identifier)
+        logging.debug("Evidence retrieval")
+        thread = threading.Thread(target=elastic_search.retrieve, args=(claims, identifier))
+        thread.start()
 
     elif next_stage == 5:
-        logging.info('Query the trained model')
-        try:
-            if STANCE_DETECTION_RESULT is None:
-                logging.error("There are no stances scores")
-                databasemanager.update_step(settings.results_table_name, settings.status, settings.error, identifier)
-                databasemanager.update_step(settings.results_table_name, settings.error_msg,
-                                            "There are no stances scores", identifier)
-            else:
-                tempjson = json.loads(STANCE_DETECTION_RESULT)
-                thread = threading.Thread(target=predictions.predict,
-                                              args=(tempjson, identifier))
-                thread.start()
-        except Exception as e:
-            logging.exception(e)
-            databasemanager.update_step(settings.results_table_name, settings.status, settings.error, identifier)
-            databasemanager.update_step(settings.results_table_name, settings.error_msg, str(e), identifier)
+        logging.debug("Stance detection")
+        thread = threading.Thread(target=cosine_similarity.calculate, args=(claims, identifier))
+        thread.start()
 
     elif next_stage == 6:
+        logging.debug('Query the trained model')
+        thread = threading.Thread(target=predictions.predict, args=(claims, identifier))
+        thread.start()
+
+    elif next_stage == 7:
+        logging.debug('Query the trained RNN model')
+        thread = threading.Thread(target=predictions.predict_mean, args=(claims, identifier))
+        thread.start()
+
+    elif next_stage == 8:
+        # run indicators if label is false
+        if veracity_label == settings.false_label:
+            indicators = run_indicator_check_text(input_text)
+            status = settings.completed
+            indicator_json = json.dumps(indicators, cls=SetEncoder)
+            databasemanager.update_json_step(settings.results_table_name, settings.results_indicator_check,
+                                        indicator_json, identifier)
+        else:
+            # otherwise show skipped status
+            status = settings.skipped
+
+        # update and increase the step
+        databasemanager.update_step(settings.results_table_name, settings.results_indicator_check_status,
+                                    status, identifier)
+        databasemanager.increase_the_stage(settings.results_table_name, identifier)
+        goNextLevel(identifier)
```

Comment:
We don't know what our partners want to do in this indicator check in the future. Since we already have the thread-based approach for the other steps, I would suggest to use it here as well. At least to me, that seems to be the saver way, just in case the indicator check takes some more time in the future. :thinking: 

On the other hand: I am wondering whether we even need all these threads that we start... Wouldn't it be sufficient to do it once at the very beginning for a newly created job? To me, it looks like one thread 1 is created, does its job, creates thread 2, terminates and than thread 2 runs and performs the next step. This is a little bit costly, when thread 1 could perform the steps sequentially :thinking: 

Maybe we should discuss this as :slightly_smiling_face:

Reply:
You are right, I've removed all the extra threads and it seems to work fine so far 🙂